trin-core/src/utp/stream.rs

```diff
@@ -855,16 +984,238 @@ impl UtpSocket {
+        } else {
+            // Find index following the most recent packet before the one we wish to insert
+            let i = self
+                .incoming_buffer
+                .iter()
+                .filter(|p| p.seq_nr() < packet.seq_nr())
+                .count();
+
+            if self
+                .incoming_buffer
+                .get(i)
+                .map_or(true, |p| p.seq_nr() != packet.seq_nr())
+            {
+                self.incoming_buffer.insert(i, packet);
+            }
+        }
+    }
+
+    /// Gracefully closes connection to peer.
+    ///
+    /// This method allows both peers to receive all packets still in
+    /// flight.
+    pub async fn close(&mut self) -> anyhow::Result<()> {
+        // Nothing to do if the socket's already closed or not connected
+        if self.state == SocketState::Closed
+            || self.state == SocketState::Uninitialized
+            || self.state == SocketState::SynSent
+        {
+            return Ok(());
+        }
+
+        // Flush unsent and unacknowledged packets
+        self.flush().await?;
+
+        let mut packet = Packet::new();
+        packet.set_connection_id(self.sender_connection_id);
+        packet.set_seq_nr(self.seq_nr);
+        packet.set_ack_nr(self.ack_nr);
+        packet.set_timestamp(now_microseconds());
+        packet.set_type(PacketType::Fin);
+
+        // Send FIN
+        self.socket
+            .send_talk_req(
+                self.connected_to.clone(),
+                ProtocolId::Utp,
+                Vec::from(packet.as_ref()),
+            )
+            .await
+            .unwrap();
+
+        debug!("CLosing connection, sent {:?}", packet);
+        self.state = SocketState::FinSent;
+
+        // Receive JAKE
+        let mut buf = [0; BUF_SIZE];
+        while self.state != SocketState::Closed {
+            self.recv(&mut buf).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Consumes acknowledgements for every pending packet.
+    pub async fn flush(&mut self) -> anyhow::Result<()> {
+        let mut buf = [0u8; BUF_SIZE];
+        while !self.send_window.is_empty() {
+            debug!("packets in send window: {}", self.send_window.len());
+            self.recv(&mut buf).await?;
+        }
+
+        Ok(())
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use crate::portalnet::discovery::Discovery;
-    use crate::portalnet::types::messages::PortalnetConfig;
+    use crate::portalnet::types::messages::{PortalnetConfig, ProtocolId};
+    use crate::portalnet::Enr;
     use crate::utils::node_id::generate_random_remote_enr;
     use crate::utp::packets::{Packet, PacketType};
     use crate::utp::stream::{SocketState, UtpSocket, BUF_SIZE};
+    use discv5::Discv5Event;
+    use std::convert::TryFrom;
+    use std::str::FromStr;
     use std::sync::Arc;
+    use tokio::sync::mpsc;
+    use tokio::sync::mpsc::UnboundedSender;
+    use tokio::sync::RwLock;
+
+    macro_rules! iotry {
```

Comment:
I'd suggest at least a comment here to explain what this means

Reply:
I'm removing this, I think we don't need it, it seems a fancy way to get rid of `unwrap` in the test code.